### ISSUE TITLE
linkerd-cni logs are flooded on some pods

### ISSUE THREAD

**Issue by relu** [linkerd/linkerd2#9916]
### What is the issue?

Some linkerd-cni pods have entered a state where the logs are continuously flooded.

### How can it be reproduced?

I have not been able to reproduce the issue, it was a one-time occurrence and only affected 3 pods out of ~90 we have running in one of our clusters. They had one thing in common: they'd been running for more than 20 days.

### Logs, error output, etc

The files on the host look like this:
```
# /host/etc/cni/net.d
01-linkerd-cni.conf  05-cilium.conflist  ZZZ-linkerd-cni-kubeconfig
```
From what I've observed, the `01-linkerd-cni.conf` file is being regenerated endlessly.

In the logs these messages are repeated over and over:
```
Created CNI config /host/etc/cni/net.d/05-cilium.conflist      
Detected change in /host/etc/cni/net.d/: CREATE 01-linkerd-cni.conf                                             
sha256sum: /host/etc/cni/net.d/01-linkerd-cni.conf: No such file or directory                                   
New file [01-linkerd-cni.conf] detected; re-installing in "chained" mode                                        
Using CNI config template from CNI_NETWORK_CONFIG environment variable.                                         
      "k8s_api_root": "https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__",                        
      "k8s_api_root": "https://172.20.0.1:__KUBERNETES_SERVICE_PORT__",                                         
CNI config: {                                                                                                   
  "name": "linkerd-cni",
  "type": "linkerd-cni",  
  "log_level": "info",
  "policy": {
      "type": "k8s",                                                                                            
      "k8s_api_root": "https://172.20.0.1:443",          
      "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"                 
  },                                                                                                            
  "kubernetes": {                                                                                               
      "kubeconfig": "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig"        
  },                                                                                                            
  "linkerd": {                                                                                                  
    "incoming-proxy-port": 4143,
    "outgoing-proxy-port": 4140,
    "proxy-uid": 2102,  
    "ports-to-redirect": [],
    "inbound-ports-to-ignore": ["4191","4190"],                                                                 
    "simulate": false,
    "use-wait-flag": false                                                                                      
  }                                                                                                             
}   
Created CNI config /host/etc/cni/net.d/01-linkerd-cni.conf                                                      
Detected change in /host/etc/cni/net.d/: DELETE 01-linkerd-cni.conf
Detected change in /host/etc/cni/net.d/: DELETE 05-cilium.conflist                                              
Detected change in /host/etc/cni/net.d/: CREATE 05-cilium.conflist                                              
New file [05-cilium.conflist] detected; re-installing in "chained" mode                                         
Using CNI config template from CNI_NETWORK_CONFIG environment variable.                                         
      "k8s_api_root": "https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__",                        
      "k8s_api_root": "https://172.20.0.1:__KUBERNETES_SERVICE_PORT__",                                         
CNI config: {                                                                                                   
  "name": "linkerd-cni",
  "type": "linkerd-cni",  
  "log_level": "info",
  "policy": {
      "type": "k8s",                                                                                            
      "k8s_api_root": "https://172.20.0.1:443",                    
      "k8s_auth_token": "__SERVICEACCOUNT_TOKEN__"                
  },                                                                                                            
  "kubernetes": {                                                                                               
      "kubeconfig": "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig"        
  },                                                                                                            
  "linkerd": {                                                                                                  
    "incoming-proxy-port": 4143,
    "outgoing-proxy-port": 4140,
    "proxy-uid": 2102,  
    "ports-to-redirect": [],
    "inbound-ports-to-ignore": ["4191","4190"],                                                                 
    "simulate": false,
    "use-wait-flag": false                                                                                      
  }                                                                                                             
}   
Removing Linkerd's configuration file: /host/etc/cni/net.d/01-linkerd-cni.conf
```

### output of `linkerd check -o short`

```
Linkerd core checks
===================

linkerd-version
---------------
‼ cli is up-to-date
    is running version 2.12.1 but the latest stable version is 2.12.2
    see https://linkerd.io/2.12/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.12.1 but the latest stable version is 2.12.2
    see https://linkerd.io/2.12/checks/#l5d-version-control for hints

linkerd-control-plane-proxy
---------------------------
‼ control plane proxies are up-to-date
    some proxies are not running the current version:
        * linkerd-destination-656d6f8b58-6vzl7 (stable-2.12.1)
        * linkerd-destination-656d6f8b58-8n2zt (stable-2.12.1)
        * linkerd-destination-656d6f8b58-9v5rj (stable-2.12.1)
        * linkerd-destination-656d6f8b58-c6tfg (stable-2.12.1)
        * linkerd-destination-656d6f8b58-m6w5d (stable-2.12.1)
        * linkerd-identity-84cbc98c59-8d8ng (stable-2.12.1)
        * linkerd-identity-84cbc98c59-kvqv9 (stable-2.12.1)
        * linkerd-identity-84cbc98c59-lcvvx (stable-2.12.1)
        * linkerd-identity-84cbc98c59-pjlsz (stable-2.12.1)
        * linkerd-identity-84cbc98c59-qhbmt (stable-2.12.1)
        * linkerd-proxy-injector-5fb97ff8c-46plj (stable-2.12.1)
        * linkerd-proxy-injector-5fb97ff8c-92mrh (stable-2.12.1)
        * linkerd-proxy-injector-5fb97ff8c-9vzmj (stable-2.12.1)
        * linkerd-proxy-injector-5fb97ff8c-d9g5c (stable-2.12.1)
        * linkerd-proxy-injector-5fb97ff8c-l8vb2 (stable-2.12.1)
    see https://linkerd.io/2.12/checks/#l5d-cp-proxy-version for hints
```

### Environment

- Kubernetes v1.23
- EKS
- Amazon Linux v2
- Linkerd v2.12.1

### Possible solution

Recycling the pods did resolve the issue. Looking at the logs, there are some repeated "Detected change" lines, I think the problem might be with the inotify monitor that gets in a state where it recursively re-triggers itself.

### Would you like to work on fixing this bug?

maybe

**Comment by mateiidavid** [comment]
@relu what version of Cilium have you installed, and did you do it through Helm or the CLI? If you can get some manifests, we might be able to try running this in a local environment. I'm curious what would happen if we take our installer script out of the equation and replace it with a dummy file.

My idea would be to run a pod that mounts the hostFS, create a placeholder conf file (e.g `01-foobar.conf`) and observe through inotify if Cilium does indeed attempt to re-create (or `mv`) files. We've fixed some issues with the inotify script so I'm not ruling it out _just yet_.

**Comment by relu** [comment]
I have this issue on Cilium 1.11.7 and 1.12.4. It's installed via Helm.

Here are the helm release values:
```
containerRuntime:
  integration: containerd
enableIPv4Masquerade: false
eni:
  enabled: true
extraConfig:
  aws-instance-limit-mapping: |
    {
      "c6i.xlarge": "4,15,15",
      "c6a.xlarge": "4,15,15",
      "m6i.xlarge": "4,15,15",
      "m6a.xlarge": "4,15,15",
      "r6i.xlarge": "4,15,15",
      "c6i.2xlarge": "4,15,15",
      "c6a.2xlarge": "4,15,15",
      "m6i.2xlarge": "4,15,15",
      "m6a.2xlarge": "4,15,15",
      "r6i.2xlarge": "4,15,15",
      "c6i.4xlarge": "8,30,30",
      "c6a.4xlarge": "8,30,30",
      "m6i.4xlarge": "8,30,30",
      "m6a.4xlarge": "8,30,30",
      "r6i.4xlarge": "8,30,30"
    }
hubble:
  relay:
    enabled: true
    resources:
      limits:
        cpu: 1
        memory: 256Mi
      requests:
        cpu: 100m
        memory: 256Mi
  ui:
    backend:
      resources:
        limits:
          cpu: 1
          memory: 512Mi
        requests:
          cpu: 250m
          memory: 512Mi
    enabled: true
    frontend:
      resources:
        limits:
          cpu: 10m
          memory: 16Mi
        requests:
          memory: 16Mi
ipam:
  mode: eni
ipv4NativeRoutingCIDR: <VPC CIDR>
kubeProxyReplacement: probe
labels: k8s:!io\.cilium\.k8s\.namespace\.labels\.kustomize\.toolkit\.fluxcd\.io
  k8s:!tags\.datadoghq\.com
nodePort:
  enableHealthCheck: false
operator:
  priorityClassName: system-cluster-critical
  prometheus:
    enabled: true
  resources:
    limits:
      cpu: 1
      memory: 256Mi
    requests:
      cpu: 50m
      memory: 256Mi
priorityClassName: system-node-critical
prometheus:
  enabled: true
resources:
  limits:
    cpu: 1
    memory: 512Mi
  requests:
    cpu: 100m
    memory: 512Mi
svcSourceRangeCheck: false
tunnel: disabled
```

I've been looking into this as well but wasn't able to pin-point exactly where and when the problem occurs. The way I've managed to reproduce it a few times was to just rollout restart the cilium agent daemonset, however, it doesn't always occur when I do that.

Something that I think would be relevant is the way Cilium takes control of the CNI config dir [in an exclusive way](https://github.com/cilium/cilium/blob/v1.12.4/plugins/cilium-cni/cni-install.sh#L134-L149). However, even if I disabled the exclusive mode I was still able to reproduce the issue so it's probably not that. [This might be more relevant](https://github.com/cilium/cilium/blob/v1.12.4/plugins/cilium-cni/cni-install.sh#L93-L104) though and could actually be the trigger (some race condition with the `find` utility?).

Please do let me know if there's anything else I can provide to help debug this. Thanks!

**Comment by alpeb** [comment]
@relu are you still having issues? I tried replicating this (using EKS with Cilium) and hit an issue we had previously experienced with Calico, that got fixed with #9504, which was included as of linkerd stable-2.12.2. So using the latest linkerd stable did the trick for me. Are you able to give this a try with stable-2.12.3 (or better yet the latest edge)?

**Comment by relu** [comment]
Thanks for following up @alpeb! I've been running on `stable-2.12.2` for some time now and I can confirm the issue is gone. I am planning to upgrade soon to `stable-2.12.3`.
